lib/exercises.rb

```diff
@@ -2,19 +2,65 @@
 
 # This method will return an array of arrays.
 # Each subarray will have strings which are anagrams of each other
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: O((n * m), where n is the number of elements in strings and m is the average length of an element in strings)
+# Space Complexity: O(n)
 
 def grouped_anagrams(strings)
-  raise NotImplementedError, "Method hasn't been implemented yet!"
+  anagram_hashes = []
+  strings.each do |string|
+    anagram_match = false
+    new_hash = {}
+    string.each_char do |n|
+      if new_hash[n]
+        new_hash[n] += 1
+      else
+        new_hash[n] = 1
+      end 
+    end
+    anagram_hashes.each do |stored_hash|
+      if stored_hash["hash"] == new_hash 
+        stored_hash["words"] << string
+        anagram_match = true
+      end 
+    end 
+    if anagram_match == false
+      anagram_hashes << {
+        "words" => [string],
+        "hash" => new_hash
+      }
+    end 
+  end 
+  list = []
+  anagram_hashes.each do |current_hash|
+    list << current_hash["words"]
+  end 
+  return list 
 end
 
 # This method will return the k most common elements
-# in the case of a tie it will select the first occuring element.
-# Time Complexity: ?
-# Space Complexity: ?
+# in the case of a tie it will select the first occurring element.
+# Time Complexity: O(n + k, where n is the length of the list parameter)
+# Space Complexity: O(n, where n is the number of unique elements in the list)
 def top_k_frequent_elements(list, k)
-  raise NotImplementedError, "Method hasn't been implemented yet!"
+  frequency_hash = {}
+  list.each do |n|
+    if frequency_hash[n] 
+      frequency_hash[n] += 1
+    else 
+      frequency_hash[n] = 1
+    end 
+  end 
+  answer = []
+  k.times do 
+    highest_freq = 0
+    frequency_hash.each_value do |value|
+      highest_freq = value if value > highest_freq
+    end 
```

Comment:
This and `frequency_hash.delete` is O(n), and since it's inside a k.times loop...